File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/reflections/LoggingArray.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 

Imports System
Imports com.db4o.reflect
Imports com.db4o.reflect.net

Namespace com.db4odoc.f1.reflections
    Public Class LoggingArray
        Implements ReflectArray
        Private ReadOnly _reflector As Reflector

        Public Sub New(ByVal pReflector As Reflector)
            _reflector = pReflector
        End Sub

        Public Function Dimensions(ByVal arr As Object) As Integer() Implements ReflectArray.Dimensions
            Return New Integer() {GetLength(arr)}
        End Function

        Function [Get](ByVal onArray As Object, ByVal index As Integer) As Object Implements ReflectArray.Get
            Return (CType(onArray, Object()))(index)
        End Function

        Function Flatten(ByVal a_shaped As Object, ByVal a_dimensions() As Integer, ByVal a_currentDimension As Integer, ByVal a_flat() As Object, ByVal a_flatElement As Integer) As Integer Implements ReflectArray.Flatten
            Dim shaped() As Object = CType(a_shaped, Object())
            System.Array.Copy(shaped, 0, a_flat, 0, shaped.Length)
            Return shaped.Length
        End Function

        Function GetComponentType(ByVal a_class As ReflectClass) As ReflectClass Implements ReflectArray.GetComponentType
            While a_class.IsArray()
                a_class = a_class.GetComponentType()
            End While
            Return a_class
        End Function

        Public Function GetLength(ByVal array As Object) As Integer Implements ReflectArray.GetLength
            Return (CType(array, Object())).Length
        End Function

        Function IsNDimensional(ByVal a_class As ReflectClass) As Boolean Implements ReflectArray.IsNDimensional
            Return False
        End Function

        Function GetNetType(ByVal a_class As ReflectClass) As Type
            Return (CType(a_class, NetClass)).GetNetType()
        End Function

        Function NewInstance(ByVal componentType As ReflectClass, ByVal dimensions As Integer()) As Object Implements ReflectArray.NewInstance
            Return NewInstance(componentType, dimensions(0))
        End Function


        Function NewInstance(ByVal componentType As ReflectClass, ByVal length As Integer) As Object Implements ReflectArray.NewInstance
            Return System.Array.CreateInstance(GetNetType(componentType), length)
        End Function


        Sub [Set](ByVal onArray As Object, ByVal index As Integer, ByVal element As Object) Implements ReflectArray.Set
            Dim shaped() As Object = CType(onArray, Object())
            shaped.SetValue(element, index)
            Return
        End Sub

        Function Shape(ByVal a_flat() As Object, ByVal a_flatElement As Integer, ByVal a_shaped As Object, ByVal a_dimensions() As Integer, ByVal a_currentDimension As Integer) As Integer Implements ReflectArray.Shape
            Dim shaped() As Object = CType(a_shaped, Object())
            System.Array.Copy(a_flat, 0, shaped, 0, a_flat.Length)
            Return a_flat.Length
        End Function
    End Class
End Namespace

File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/reflections/LoggingArray.cs
/* Copyright (C) 2005   db4objects Inc.   http://www.db4o.com */

using System;
using com.db4o.reflect;
using com.db4o.reflect.net;

namespace com.db4odoc.f1.reflections
{
	public class LoggingArray : com.db4o.reflect.ReflectArray
	{
		private readonly com.db4o.reflect.Reflector _reflector;

		internal LoggingArray(com.db4o.reflect.Reflector reflector)
		{
			_reflector = reflector;
		}

		public virtual int[] Dimensions(object arr)
		{
			return new int[] { GetLength(arr) };
		}

		public virtual int Flatten(object a_shaped, int[] a_dimensions, int a_currentDimension
			, object[] a_flat, int a_flatElement)
		{
			object[] shaped = (object[])a_shaped;
			System.Array.Copy(shaped, 0, a_flat, 0, shaped.Length);
			return shaped.Length;
		}

		public virtual object Get(object onArray, int index)
		{
			return ((object[])onArray)[index];
		}

		public virtual com.db4o.reflect.ReflectClass GetComponentType(com.db4o.reflect.ReflectClass
			a_class)
		{
			while (a_class.IsArray()) 
			{
				a_class = a_class.GetComponentType();
			}
			return a_class;
		}

		public virtual int GetLength(object array)
		{
			return ((object[])array).Length;
		}

		public virtual bool IsNDimensional(com.db4o.reflect.ReflectClass a_class)
		{
			return false;
		}

		private static Type GetNetType(ReflectClass a_class)
		{
			return ((NetClass)a_class).GetNetType();
		}

		public virtual object NewInstance(com.db4o.reflect.ReflectClass componentType, int
			length)
		{
			return System.Array.CreateInstance(GetNetType(componentType), length);
		}

		public virtual object NewInstance(com.db4o.reflect.ReflectClass componentType, int[]
			dimensions)
		{
			return NewInstance(componentType, dimensions[0]);
		}

		public virtual void Set(object onArray, int index, object element)
		{
			((object[])onArray)[index] = element;
			return;
		}

		public virtual int Shape(object[] a_flat, int a_flatElement, object a_shaped, int[]
			a_dimensions, int a_currentDimension)
		{
			object[] shaped = (object[])a_shaped;
			System.Array.Copy(a_flat, 0, shaped, 0, a_flat.Length);
			return a_flat.Length;
		}
	}
}

